fetch.bs

```diff
@@ -6615,6 +6706,62 @@ agent's <a>CORS-preflight cache</a> for which there is a <a>cache entry match</a
 </ol>
 </div>
 
+<h3 id=deferred-fetching>Deferred fetching</h3>
+
+<p>Deferred fetches allow callers to request that a fetch is invoked at the latest possible moment,
+when a <a>fetch group</a> is <a for="fetch group">terminated</a>, or after a timeout after it is
+<a for="fetch group">deactivated</a>.
+
+<div algorithm="request-a-deferred-fetch">
+<p>To <dfn>request a deferred fetch</dfn> given a
+<a for=/>request</a> <var>request</var> and a null-or-{{DOMHighResTimeStamp}}
+<var>inactivityDeferredDelay</var> (default null):
+
+<ol>
+ <li><p>Assert: <var>request</var>'s <a for=request>client</a> is an
+ <a>environment settings object</a>.
+
+ <li>Let <var>totalScheduledDeferredBytesForOrigin</var> be 0.
+
+ <li>
+  <p>If <var>request</var>'s <a for=request>body</a> is not null then:
+
+  <ol>
+   <li><p>If <var>request</var>'s
+   <a for=request>body</a>'s <a for=body>length</a> is null, then throw a {{TypeError}}.
+
+   <li><p>Set <var>totalScheduledDeferredBytesForOrigin</var> to <var>request</var>'s
+   <a for=request>body</a>'s <a for=body>length</a>.
+  </ol>
+ </li>
+
+ <li><p><a for=list>For each</a> <a>deferred fetch record</a> <var>deferredRecord</var> in
+ <var>request</var>'s <a for=request>client</a>'s <a for=fetch>fetch group</a>'s
+ <a for="fetch group">deferred fetch records</a>: if <var>deferredRecord</var>'s
+ <a for="deferred fetch record">request</a>'s <a for=request>body</a> is not null and
+ <var>deferredRecord</var>'s <a for="deferred fetch record">request</a>'s <a for=request>URL</a>'s
+ <a for=url>origin</a> is <a>same origin</a> with <var>request</var>'s <a for=request>URL</a>'s
+ <a for=url>origin</a>, then increment <var>totalScheduledDeferredBytesForOrigin</var> by
+ <var>deferredRecord</var>'s <a for="deferred fetch record">request</a>'s <a for=request>body</a>'s
+ <a for=body>length</a>.
+
+ <li><p>If <var>totalScheduledDeferredBytesForOrigin</var> is greater than 64 kilobytes, then
+ throw a {{QuotaExceededError}}.
```

Comment:
If the discussion at https://github.com/whatwg/fetch/pull/1647#discussion_r1197485829 is resolved and the API shape is changed to return a `DeferredFetchResult` instead of `Promise`, should we consider to return null here in addition to throw an error?

Reply:
There's no return value when you throw.

Reply:
I understand that the body length case needs to fail but I thought we were going to flush the existing entries in this case. Why are we throwing in that case?

Reply:
We've discussed this. The body-length case has a per-target-origin quota, It's up to userland to flush if it's reached, same as with `keepalive`.

Reply:
We have a issue (https://github.com/WICG/pending-beacon/issues/76) tracking this topic, and I don't remember about the decision to hand flushing off to userland implementation.

Also, I thought the quota of `keepalive` is specified by [spec](https://www.w3.org/TR/beacon/#sec-processing-model), and implemented by UA (not userland JS code)?

```
All requests with this flag set share the same in-flight quota restrictions that is enforced within the Fetch API.
```

Reply:
I know we had discussions about this but I don't remember coming to this conclusion (or really any strong conclusion). I think it's very difficult to manage this in userland. Beacons can be created in the 1st party origin by any 3rd party script. Even keeping it all 1st party, it would require *all* beacon usage in all 1st-party frames to go through a central point in order to correctly account for the bytes. That seems like a heavy burden on devs.

I realise that speccing this behaviour might be hard and maybe could come later. Implementers have some leeway on when to send, so maybe we could implement the autoflushing behaviour even if it's not explicitly in the spec.

Reply:
We talked about having the quota for deferred fetches based on the 3p origin, so each origin manages its own and autoflushes. Otherwise we have to start managing the "sent" state differently because those fetches can be flushed at any time.

Reply:
> Keepalive and fetchLater are mutually exclusive and so are there quotas.

> For background-timeout fetches, the UA can dispatch them at any time to make sure they don't surpass the keepalive quota.

Could you help clarify which quota `fetchLater()` requests are actually bound to? Does specifying `backgroundTimeout` change the quota?

Reply:
See how `totalScheduledDeferredBytesForOrigin` works in [request a deferred fetch](https://whatpr.org/fetch/1647.html#request-a-deferred-fetch). Specifying `backgroundTiming` doesn't change the quota

Reply:
@mingyc anything else here or can we resolve this?

Reply:
Thanks for explaining. I added some more questions below

> in  request’s [ client ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-request-client)’s [ fetch group ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-fetch-group) ’s [ deferred fetch records ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-defer=fetch-record): ...
> if  deferredRecord’s [ request ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#deferred-fetch-record-request)’s [ body ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-request-body) is not null and  deferredRecord’s [ request ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#deferred-fetch-record-request)’s [ URL ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-request-url)’s [ origin ](https://url.spec.whatwg.org/#concept-url-origin) is [ same origin ](https://html.spec.whatwg.org/multipage/browsers.html#same-origin) with  request’s [ URL ](https://whatpr.org/fetch/1647/094ea69...02a45de.html#concept-request-url)’s [ origin ](https://url.spec.whatwg.org/#concept-url-origin) ...
> If  `totalScheduledDeferredBytesForOrigin`  is greater than 64 kilobytes, then throw a [ QuotaExceededError ](https://webidl.spec.whatwg.org/#quotaexceedederror).

It means `totalScheduledDeferredBytesForOrigin` only accumulates request lenth from same-origin deferred fetch within the same fetch group (e.g. same Document).

1. You mentioned that the quota is `per-target-origin quota`, how will this work with  records from different fetch group?
2. `each origin manages its own and autoflushes` is not specified in above. As Fergal mentioned, it would be difficult for userland code to track request size and flush by themselves. Could we specify auto flushes in FIFO order?

Reply:
1. per-target-origin-in-document (a fetch group is a document)
2. Autoflushing would require rewriting this whole thing as suddenly you could have `committed` in a live document. I don't see why tit would be difficult to do this in userland. You can catch `QuotaExceededError` from `fetchLater` and perform a regular fetch.

Reply:
The discussions about Promise/autoflushes are resolved.

However, some questions on quota are not yet:

> Regarding keepalive, I think this should be a bit different (also chatted about this with @yoavweiss):
Keepalive and fetchLater are mutually exclusive and so are there quotas.
>
> Keepalive quota is global, fetchLater quota is per origin
> Keepalive is for requests that dispatch immediately
> Keepalive survives termination, fetchLater (usually) invokes at termination
> 
> So I think the solution would be that deferred fetches are not keepalive, and instead, they're simply not terminated when the fetch group is terminated,

@noamr We now have fetchLater-specific quota steps, but fetchLater requests are also set keepalive true. Should fetchLater requests share the quota of non-FetchLater keepalive requests, i.e. should the former also be run through the same inflight keepalive checks?

Reply:
We've talked about this before in this long issue discussion :)
- These request are implicitly keepalive, because they're never registered as fetch records. we don't actually set their keepalive to true in other ways.
- There is no point in sharing the regular keepalive quota. It would mean that origins would fight for that quota, and it is also unnecessary because the UA can simply delay the deferred fetches until all the keepalive fetches are complete. In addition the caller wouldn't receive an error for this which would make `fetchLater` unreliable.